core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala

```diff
@@ -77,5 +80,124 @@ class BrokerMetadataListenerTest {
+          false
+        }
+      } catch {
+        case t: Throwable => failure.compareAndSet(null, t)
+      }
+    }
+  }
+
+  class MockMetadataPublisher extends MetadataPublisher {
+    var image = MetadataImage.EMPTY
+
+    override def publish(newHighestMetadataOffset: Long,
+                         delta: MetadataDelta,
+                         newImage: MetadataImage): Unit = {
+      image = newImage
+    }
+  }
+
+  private val FOO_ID = Uuid.fromString("jj1G9utnTuCegi_gpnRgYw")
+
+  private def generateManyRecords(listener: BrokerMetadataListener,
+                                  endOffset: Long): Unit = {
+    (0 to 10000).foreach { _ =>
+      listener.handleCommit(RecordTestUtils.mockBatchReader(endOffset,
+        util.Arrays.asList(new ApiMessageAndVersion(new PartitionChangeRecord().
+          setPartitionId(0).
+          setTopicId(FOO_ID).
+          setRemovingReplicas(Collections.singletonList(1)), 0.toShort),
+          new ApiMessageAndVersion(new PartitionChangeRecord().
+            setPartitionId(0).
+            setTopicId(FOO_ID).
+            setRemovingReplicas(Collections.emptyList()), 0.toShort))))
+    }
+    listener.getImageRecords().get()
+  }
+
+  @Test
+  def testCreateSnapshot(): Unit = {
+    val snapshotter = new MockMetadataSnapshotter()
+    val listener = new BrokerMetadataListener(0, Time.SYSTEM, None, 1000L, snapshotter)
+    try {
+      (0 to 3).foreach {
+        id => listener.handleCommit(RecordTestUtils.mockBatchReader(100L,
+            util.Arrays.asList(new ApiMessageAndVersion(new RegisterBrokerRecord().
+              setBrokerId(id).
+              setBrokerEpoch(100L).
+              setFenced(false).
+              setRack(null).
+              setIncarnationId(Uuid.fromString("GFBwlTcpQUuLYQ2ig05CS" + id)), 1))))
+      }
+      listener.handleCommit(RecordTestUtils.mockBatchReader(200L,
+        util.Arrays.asList(new ApiMessageAndVersion(new TopicRecord().
+            setName("foo").
+            setTopicId(FOO_ID), 1.toShort),
+          new ApiMessageAndVersion(new PartitionRecord().
+            setPartitionId(0).
+            setTopicId(FOO_ID).
+            setIsr(util.Arrays.asList(0, 1, 2)).
+            setLeader(0).
+            setReplicas(util.Arrays.asList(0, 1, 2)).
+            setRemovingReplicas(util.Arrays.asList(0, 1, 2)).
+            setAddingReplicas(util.Arrays.asList(0, 1, 2)), 1.toShort))))
+      listener.getImageRecords().get()
+      assertEquals(200L, listener.highestMetadataOffset())
+
+      // Check that we generate at least one snapshot once we see enough records.
+      assertEquals(-1L, snapshotter.prevCommittedOffset)
+      generateManyRecords(listener, 1000L);
+      assertEquals(1000L, snapshotter.prevCommittedOffset)
+      assertEquals(1000L, snapshotter.activeSnapshotOffset)
+      snapshotter.activeSnapshotOffset = -1L
+
+      // Test creating a new snapshot after publishing it.
+      val publisher = new MockMetadataPublisher()
```

Comment:
I'm probably missing something, but it's not clear to me why publishing makes this test case interesting.

Reply:
Publishing does involve a slightly different code path (although not that different).

For one thing, once we're publishing, the delta will be empty when `_delta.apply()` is called (which is fine, it just means we get the image the delta is based on, unmodified).

Anyway, it's an easy test to do, and quick as well, so I thought it was worth it here.